### PR TITLE
[SPARK-20196][PYTHON][SQL] update doc for catalog functions for all languages, add pyspark refreshByPath API

### PR DESCRIPTION
**Pull request by felixcheung**
## What changes were proposed in this pull request?

Update doc to remove external for createTable, add refreshByPath in python

## How was this patch tested?

manual

### CODE REVIEW

**Comment by felixcheung** [comment]
@gatorsmile

**Comment by felixcheung** [comment]
will update after #17518 + changes to R doc too

**Comment by felixcheung** [comment]
updated @gatorsmile

File: R/pkg/R/SQLContext.R [hunk 1]
```diff
@@ -544,12 +544,15 @@ sql <- function(x, ...) {
   dispatchFunc("sql(sqlQuery)", x, ...)
 }
 
-#' Create a SparkDataFrame from a SparkSQL Table
+#' Create a SparkDataFrame from a SparkSQL table or temporary view
```
Comment (gatorsmile) [inline R/pkg/R/SQLContext.R]: `table or view`

Here, actually, it includes both temporary views or persistent views.

File: R/pkg/R/SQLContext.R [hunk 2]
```diff
@@ -544,12 +544,15 @@ sql <- function(x, ...) {
...
 #'
-#' Returns the specified Table as a SparkDataFrame.  The Table must have already been registered
-#' in the SparkSession.
+#' Returns the specified table or temporary view as a SparkDataFrame. The temporary view must have
```
Comment (gatorsmile) [inline R/pkg/R/SQLContext.R]: The same here.

File: python/pyspark/sql/catalog.py [hunk 3]
```diff
@@ -276,14 +277,24 @@ def clearCache(self):
...
-        """Invalidate and refresh all the cached metadata of the given table."""
+        """Invalidates and refreshes all the cached data and metadata of the given table."""
         self._jcatalog.refreshTable(tableName)
 
     @since('2.1.1')
     def recoverPartitions(self, tableName):
-        """Recover all the partitions of the given table and update the catalog."""
+        """Recovers all the partitions of the given table and update the catalog.
+
+        Only works with a partitioned table, and not a temporary view.
```
Comment (gatorsmile) [inline python/pyspark/sql/catalog.py]: `a temporary view` -> `a view`

Both temporary and persistent views are not supported. We will detect and issue exceptions.

File: sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala [hunk 4]
```diff
@@ -447,6 +447,7 @@ abstract class Catalog {
 
   /**
    * Recovers all the partitions in the directory of a table and update the catalog.
+   * Only works with a partitioned table, and not a temporary view.
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/catalog/Catalog.scala]: The same here.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala [hunk 5]
```diff
@@ -456,7 +460,8 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
 
   /**
    * Refreshes the cache entry and the associated metadata for all Dataset (if any), that contain
-   * the given data source path.
+   * the given data source path. Path matching is by prefix, i.e. "/" would invalidate
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: `invalidate` -> `refresh`

We also do the re-cache.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala [hunk 6]
```diff
@@ -367,6 +367,7 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
 
   /**
    * Recovers all the partitions in the directory of a table and update the catalog.
+   * Only works with a partitioned table, and not a temporary view.
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: The same here.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala (hunk 5 above)
Comment (felixcheung) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: For some reason in here, CatalogImpl.scala is very different from Catalog.scala - let me know if you want me to change them - for now I've updated the first sentence.

**Comment by felixcheung** [comment]
Jenkins, retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala (hunk 6 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: `not a temporary view.`
-> 
`not a view.`

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala (hunk 5 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: Yes. I found this sentence is copied from Catalog.scala. Maybe, we can update them to
> Path matching is by prefix, i.e. "/" would invalidate all the cached entries and make the new versions cached lazily.

File: R/pkg/R/SQLContext.R [hunk 7]
```diff
@@ -544,12 +544,15 @@ sql <- function(x, ...) {
   dispatchFunc("sql(sqlQuery)", x, ...)
 }
 
-#' Create a SparkDataFrame from a SparkSQL Table
+#' Create a SparkDataFrame from a SparkSQL table or view
 #'
-#' Returns the specified Table as a SparkDataFrame.  The Table must have already been registered
-#' in the SparkSession.
+#' Returns the specified table or view as a SparkDataFrame. The table or view must already exists or
```
Comment (gatorsmile) [inline R/pkg/R/SQLContext.R]: Nit: `exists` -> `exist`

**Comment by gatorsmile** [comment]
LGTM except minor comments.

File: R/pkg/R/SQLContext.R (hunk 7 above)
Comment (felixcheung) [inline R/pkg/R/SQLContext.R]: fixed, thanks for catching this!

**Comment by felixcheung** [comment]
merged to master, thanks!

**Pull request closed without merging**